File: src/statemachine_pulser.c

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <unistd.h>
#include "pr_utils.h" 
#include "cce_1_def.h" 


/*

gcc -o -lpigpio -lpthread

program to readout the counters 

*/


int main(int argc, char *argv[]){
	int counts[6];
   int ls[8];// read status 
   printf("start program %s\n\r", argv[0]);
    // init hardware 
	int err= init_cce_2();
	if (err) {
		printf("initialization of hardware failed with err %d \n\r", err);
		return err; 
	}
	setup_output_array( Dpins , 8 );
	
	const int datavalid = GPIO15;
	gpioSetMode(datavalid,PI_OUTPUT); 
	gpioWrite (datavalid , 0) ;	 
	
	const int start= GPIO14;
	gpioSetMode(start,PI_OUTPUT); 
	gpioWrite (start, 0) ;	
	printf( "set start == GPIO%d to %d \n\r",start ,gpioRead(start));
	const int ready= 22;  //gpio22
	gpioSetMode(ready,PI_OUTPUT); 
	gpioWrite (ready, 0) ;	
	
	// load the values 
    set_muxout2( 0,4); // Pulse time 
	set_outputs(Dpins, 8 , 4) ;
	gpioWrite (datavalid , 1) ;gpioWrite (datavalid , 0) ;	 //fill register 
    set_muxout2( 1,4); // wait time 
	set_outputs(Dpins, 8 , 100) ;
	gpioWrite (datavalid , 1) ;gpioWrite (datavalid , 0) ;	 //fill register 
    set_muxout2( 2,4); // Nr pulse 
	set_outputs(Dpins, 8 , 5) ;
	gpioWrite (datavalid , 1) ;gpioWrite (datavalid , 0) ;	 //fill register

	gpioWrite ( start,1) ;
	printf( "set start == GPIO%d to %d \n\r",start ,gpioRead(start));
	gpioWrite ( start,0) ;
	printf( "set start == GPIO%d to %d \n\r",start ,gpioRead(start));
	sleep( 10);
	gpioWrite ( ready,1) ;
	printf( "set ready == GPIO%d to %d \n\r",ready ,gpioRead(ready));
	gpioWrite ( ready,0) ;
	printf( "set ready == GPIO%d to %d \n\r",ready ,gpioRead(ready));
	
   gpioTerminate();
}

File: python/statemachine_pulser.py
import time
from upctrl_pinfunctions import *
#from pr_utils import ar2decvalue


def main(argv):
    
    print(f"start program {argv[0]}\n\r")
    
    # init hardware
    pi,err = init_cce_2()
    if err:
        print(f"initialization of hardware failed with err {err} \n\r")
        return err

    setup_output_array(Dpins, 8)

    datavalid = GPIO15
    pi.set_mode(datavalid, piwrap.OUTPUT)
    pi.write(datavalid, 0)

    start = GPIO14
    pi.set_mode(start, piwrap.OUTPUT)
    pi.write(start, 0)
    print(f"set start == GPIO{start} to {pi.read(start)} \n\r")

    ready = GPIO22
    pi.set_mode(ready, piwrap.OUTPUT)
    pi.write(ready, 0)

    # load the values
    set_muxout2(0, 4)  # Pulse time
    set_outputs(Dpins, 8, 4)
    pi.write(datavalid, 1)
    pi.write(datavalid, 0)  # fill register

    set_muxout2(1, 4)  # wait time
    set_outputs(Dpins, 8, 10)
    pi.write(datavalid, 1)
    pi.write(datavalid, 0)  # fill register

    set_muxout2(2, 4)  # Nr pulse
    set_outputs(Dpins, 8, 7)
    pi.write(datavalid, 1)
    pi.write(datavalid, 0)  # fill register
    
    set_muxout2(3, 4)  # spare reg
    set_outputs(Dpins, 8, 8)
    pi.write(datavalid, 1)
    pi.write(datavalid, 0)  # fill register

    pi.write(start, 1)
    print(f"set start == GPIO{start} to {pi.read(start)} \n\r")
    pi.write(start, 0)
    print(f"set start == GPIO{start} to {pi.read(start)} \n\r")

    time.sleep(10)

    pi.write(ready, 1)
    print(f"set ready == GPIO{ready} to {pi.read(ready)} \n\r")
    pi.write(ready, 0)
    print(f"set ready == GPIO{ready} to {pi.read(ready)} \n\r")

    pi.stop()

if __name__ == "__main__":
    import sys
    main(sys.argv)
